exporter/sumologicexporter/exporter_test.go

```diff
@@ -0,0 +1,279 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sumologicexporter
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type test struct {
+	srv *httptest.Server
+	exp component.LogsExporter
+	se  *sumologicexporter
+}
+
+func getExporter(t *testing.T, cb []func(req *http.Request)) *test {
+	reqCounter := 0
+	// generate a test server so we can capture and inspect the request
+	testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		res.WriteHeader(200)
+		res.Write([]byte(""))
+		if len(cb) > 0 && assert.Greater(t, len(cb), reqCounter) {
+			cb[reqCounter](req)
+			reqCounter++
+		}
+	}))
+
+	cfg := &Config{
+		URL:                testServer.URL,
+		LogFormat:          "text",
+		Client:             "otelcol",
+		MaxRequestBodySize: 20_971_520,
+	}
+	factory := NewFactory()
+	exp, err := factory.CreateLogsExporter(context.Background(), component.ExporterCreateParams{Logger: zap.NewNop()}, cfg)
+
+	assert.NoError(t, err)
+
+	return &test{
+		exp: exp,
+		srv: testServer,
+		se: &sumologicexporter{
+			config: cfg,
+		},
+	}
+}
+
+func TestGetMetadata(t *testing.T) {
+	attributes := pdata.NewAttributeMap()
+	attributes.InsertString("key3", "value3")
+	attributes.InsertString("key1", "value1")
+	attributes.InsertString("key2", "value2")
+	attributes.InsertString("additional_key2", "value2")
+	attributes.InsertString("additional_key3", "value3")
+
+	test := getExporter(t, []func(req *http.Request){func(req *http.Request) {}})
+	test.se.config.MetadataFields = []string{"^key[12]", "^key3"}
+	test.se.refreshMetadataRegexes()
+
+	metadata := test.se.GetMetadata(attributes)
+	expected := "key1=value1, key2=value2, key3=value3"
+	assert.Equal(t, expected, metadata)
+}
+
+func TestFilterOutMetadata(t *testing.T) {
+	attributes := pdata.NewAttributeMap()
+	attributes.InsertString("key3", "value3")
+	attributes.InsertString("key1", "value1")
+	attributes.InsertString("key2", "value2")
+	attributes.InsertString("additional_key2", "value2")
+	attributes.InsertString("additional_key3", "value3")
+
+	test := getExporter(t, []func(req *http.Request){func(req *http.Request) {}})
+	test.se.config.MetadataFields = []string{"^key[12]", "^key3"}
+	test.se.refreshMetadataRegexes()
+
+	data := test.se.filterMetadata(attributes, true)
+	expected := map[string]string{
+		"additional_key2": "value2",
+		"additional_key3": "value3",
+	}
+	assert.Equal(t, data, expected)
+}
+
+func extractBody(req *http.Request) string {
+	buf := new(strings.Builder)
+	io.Copy(buf, req.Body)
```

Comment:
* This returns `(written int64, err error)`
* I'd personally use [`ioutil.ReadAll`](https://golang.org/pkg/io/ioutil/#ReadAll)

Reply:
What about this? https://haisum.github.io/2017/09/11/golang-ioutil-readall/

Reply:
Yeah, good point 👍🏻  but since this is in test it should be fine.